### ISSUE TITLE
Connection pooling

### ISSUE THREAD

**Issue by nuromirg** [jsp-lqk/metapipe-memcached#5]
Hey! Awesome project, thank you for the very useful solution for meta protocols in memcache. I might have an idea of how you can potentially improve your performance and resource management. With connection pooling for your inner client, each `MemcacheClient` can manage its connections more efficiently by reducing the overhead of establishing and tearing down connections frequently. It also can help with concurrency handling too, since it will reuse existing connections spawned before (and we can tinker it with some limitations of the maximum number of connections). It also should help optimize a little more the usage of sockets and fds (if any), especially in high-load conditions

```go
type ConnectionPool struct {
	pool   chan *BaseTCPClient
	target ConnectionTarget
}

type InnerMetaClient struct {
	readPool     *ConnectionPool
	mutationPool *ConnectionPool
	target       ConnectionTarget
}
```

**Comment by jsp-lqk** [comment]
Hey, thanks for taking a look at my library!

Have you faced any performance issues? 

The library is designed to avoid connection pools by using protocol pipelining. This means holding one single connection (they should not be established and tore down frequently, there should be one single long lived connection). The inner client accepts commands and dispatches them, and stores the response channel on a deque. And whenever a response from memcached arrives, it finds the pending request on the other edge of the deque, and replies to the channel. This is a very similar approach to the one used by the [folsom memcached client](https://github.com/spotify/folsom). It also means that I have avoided connection pools on purpose :)

I have done some preliminary benchmarking, and I've found that this library, using pipelining on single connection is way more efficient than gomemcache using connection pooling: less operations, less allocations, lower latency, more scalable under heavy duress.

So from my testing, both performance and resource management in this library is better than connection pool based ones, so I must reject your suggestion, even though I truly appreciate that you took the time to read the code and send me suggestion!
